README.md

```diff
@@ -1,2 +1,331 @@
+    3. [Recovery](#33-recovery)
+4. [Replication](#4replication)
+    1. [Signers](#41-signers)
+    2. [Casts](#42-casts)
+    3. [Actions](#43-actions)
+    4. [Verifications](#44-verifications)
+    5. [Metadata](#45-metadata)
+    6. [Root Revocations](#46-root-revocations)
+    7. [Sharding](#47-sharding)
+5. [Peering](#5-peering)
+6. [Upgradeability](#6-upgradeability)
+    1. [Minor Upgrades](#61-minor-upgrades)
+    2. [Major Upgrades](#62-major-upgrades)
+7. [Security Considerations](#7-security-considerations)
+    1. [Signer Compromise](#71-signer-compromise)
+    2. [Eclipse Attacks](#72-eclipse-attacks)
+    3. [Flooding Attacks](#73-flooding-attacks)
+    4. [DDOS Attacks](#74-ddos-attacks)
+    5. [Replay Attacks](#75-replay-attacks)
+8. [URIs](#8-uris)
+9. [Governance](#9-governance)
+
+# 1. Introduction
+
+Social networks have become an essential part of our lives over the last decade. Many of them began their journey as open platforms, courting developers to build on their APIs. These developers created new clients, discovered new UI paradigms, and even launched multi-billion dollar businesses that brought in many users. But over the last few years networks have turned away from developers. They've restricted APIs, implemented arbitrary review processes and have removed access with little notice or recourse. 
+
+Farcaster is a [sufficiently decentralized](https://www.varunsrinivasan.com/2022/01/11/sufficient-decentralization-for-social-networks) protocol that empowers developers to build novel social networks. We define a sufficiently decentralized network as one where **two users who want to communicate are always able to, even if the network wants to prevent it**. Users on such networks must have full control over their identity (usernames), data (messages) and social graph (relationships to others). If a third party controls any of these, they can prevent two users from communicating. Developers must also be free to build applications and have unrestricted access to the network and users must be free to switch between them. If there was only one app to connect to the network, it could prevent two users from communicating.  
+
+# 2. Concepts
+
+Farcaster achieves sufficient decentralization through a hybrid architecture that has on-chain and off-chain components. 
+
+User identities are kept on-chain because they are valuable assets, and we want to leverage the security, composability and strong consistency guarantees of the Ethereum blockchain. On-chain identities are controlled by an Ethereum address, which is a public private key pair that can also sign off-chain messages.  
+
+User data cannot be practically stored on the Ethereum blockchain or any other L1 because of the strong privacy and scalability requirements. Instead users store their data off-chain on a server under their control known as a Farcaster Hub. All data must be cryptographically signed by the user's identity address before being published to the Hub. 
```

Comment:
I would emphasize that every user can chose what hub they use to store their data (here it sounds like users need to self-host), for instance:

"Instead users store their data off-chain on servers that we call Farcaster Hubs. Each user can choose which Farcaster Hub stores their data, and migrate from one Hub to another whenever desired."

Reply:
As a follow-up, can data only be on one hub? What about replication guarantees?